### PR TITLE
rdap_types cleanups

### PR DESCRIPTION
**Pull request by djc**

### CODE REVIEW

File: rdap_types/src/lib.rs [hunk 1]
```diff
@@ -871,79 +864,96 @@ pub struct FredNsset {
...
 
 /// https://tools.ietf.org/html/rfc7483#section-5.3
-#[skip_serializing_none]
 #[derive(Serialize, Deserialize, Debug)]
 #[serde(rename_all = "camelCase")]
 pub struct Domain {
+    #[serde(skip_serializing_if = "Option::is_none")]
     pub handle: Option<String>,
+    pub ldh_name: Option<String>,
+    pub unicode_name: Option<String>,
```
Comment (valkum) [inline rdap_types/src/lib.rs]: I think these two need a `#[serde(skip_serializing_if = "Option::is_none")]` as well. The RDAP spec is rather vague about `null` vs absent members.

File: rdap_types/src/lib.rs [hunk 2]
```diff
@@ -871,79 +864,96 @@ pub struct FredNsset {
...
+    #[serde(skip_serializing_if = "Option::is_none")]
     pub nameservers: Option<Vec<Nameserver>>,
-    #[serde(rename = "secureDNS")]
+    #[serde(rename = "secureDNS", skip_serializing_if = "Option::is_none")]
     pub secure_dns: Option<SecureDns>,
-    pub remarks: Option<NoticesOrRemarks>,
-    pub events: Events,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub remarks: Option<Vec<NoticeOrRemark>>,
+    pub events: Vec<Event>,
```
Comment (valkum) [inline rdap_types/src/lib.rs:897]: Do we want a `#[serde(skip_serializing_if = "Vec::is_empty")]` here as well?

File: rdap_types/src/lib.rs [hunk 3]
```diff
@@ -871,79 +864,96 @@ pub struct FredNsset {
...
-    pub notices: Option<NoticesOrRemarks>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub notices: Option<Vec<NoticeOrRemark>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
     pub port43: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
     pub status: Option<Vec<Status>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
     pub lang: Option<String>,
     pub object_class_name: String,
```
Comment (valkum) [inline rdap_types/src/lib.rs]: Do we want to model this as a tagged enum instead? Would eliminate the case where you are able to create a Domain type with a wrong object_class_name?

File: rdap_types/src/lib.rs [hunk 4]
```diff
@@ -580,41 +592,20 @@ impl Serialize for EventAction {
...
-
-impl<'a> IntoIterator for &'a Events {
-    type Item = &'a Event;
-    type IntoIter = Iter<'a, Event>;
-
-    fn into_iter(self) -> Self::IntoIter {
-        self.0.iter()
-    }
-}
-
```
Comment (valkum) [inline rdap_types/src/lib.rs:615]: I assume this is used in the client part (or rather by users of the rdap_client)? If we use this crate just for the types, would it make sense to drop the client crate?

File: rdap_types/src/lib.rs [hunk 5]
```diff
@@ -123,28 +120,30 @@ impl<'de> Deserialize<'de> for CountryCode {
...
 /// Value signifying the relationship an object would have with its closest containing object.
 /// Values come from [RFC 7483] and [RDAP JSON Values].
 ///
 /// [RFC 7483]: https://tools.ietf.org/html/rfc7483#section-10.2.4
 /// [RDAP JSON Values]: https://www.iana.org/assignments/rdap-json-values/rdap-json-values.xhtml
-#[skip_serializing_none]
 #[derive(Serialize, Deserialize, Debug, PartialEq, Clone, Copy)]
-#[serde(rename_all = "lowercase")]
-#[serde(remote = "Role")]
+#[serde(rename_all = "lowercase", remote = "Role")]
```
Comment (valkum) [inline rdap_types/src/lib.rs:148]: I assume, the `remote` here exists to be able to implement the De/Serialize traits manually. But, the behavior of `.tirol` is actually a spec violation. If we want to maintain this, we need to figure out how we want to support these cases.
Maybe have all registered roles here and then a `Unregistered(String)` variant?

Based on: https://rdap.ryce-rsp.com/rdap/domain/nic.tirol
.tirol claims to implement both "icann_rdap_technical_implementation_guide_0" and "icann_rdap_response_profile_0". 
The technical guide says under 4.2:
> If the server responses include roles which are not listed below, such roles MUST
be registered at the IANA RDAP JSON Values Registry..

I checked the registry and `tech` and `admin` are not registered.

File: rdap_types/src/lib.rs [hunk 6]
```diff
@@ -123,28 +120,30 @@ impl<'de> Deserialize<'de> for CountryCode {
 }
 
 /// https://tools.ietf.org/html/rfc7483#section-4.2
-#[skip_serializing_none]
 #[derive(Serialize, Deserialize, Debug)]
 pub struct Link {
+    #[serde(skip_serializing_if = "Option::is_none")]
     pub value: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
     pub rel: Option<String>,
```
Comment (valkum) [inline rdap_types/src/lib.rs]: These two are actually required per: https://www.rfc-editor.org/rfc/rfc9083.html#section-4.2-4

Complete changes from RFC7483
https://www.rfc-editor.org/rfc/rfc9083.html#name-changes-from-rfc-7483

File: rdap_types/src/lib.rs [hunk 7]
```diff
@@ -666,48 +657,20 @@ impl Serialize for NoticeOrRemarkType {
...
 #[derive(Serialize, Deserialize, Debug)]
 pub struct NoticeOrRemark {
+    #[serde(skip_serializing_if = "Option::is_none")]
     pub title: Option<String>,
-    #[serde(rename = "type")]
-    pub typ: Option<NoticeOrRemarkType>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub r#type: Option<NoticeOrRemarkType>,
+    #[serde(skip_serializing_if = "Option::is_none")]
     pub description: Option<Vec<String>>,
```
Comment (valkum) [inline rdap_types/src/lib.rs:681]: This is also required based on https://www.rfc-editor.org/rfc/rfc9083.html#section-4.3-2

File: rdap_types/src/lib.rs (hunk 4 above)
Comment (djc) [reply rdap_types/src/lib.rs:615]: It's not used in the rdap_client crate. IMO it's small/trivial enough that users can pretty easily implement it themselves.

File: rdap_types/src/lib.rs (hunk 1 above)
Comment (djc) [reply rdap_types/src/lib.rs]: Right, added these. I don't think emitting explicit `null`s will cause issues for clients, I see this more as an optimization.

File: rdap_types/src/lib.rs (hunk 2 above)
Comment (djc) [reply rdap_types/src/lib.rs:897]: IMO assuming `[]` is equivalent to leaving the field out seems a bit more like a stretch, so I don't think we should do that. I contemplated making some of the `Option<Vec<_>>` fields just `Vec<_>` with `skip_serializing_if = "Vec::is_empty"` but I figured it's okay for now.

File: rdap_types/src/lib.rs (hunk 6 above)
Comment (djc) [reply rdap_types/src/lib.rs]: I left them optional for now to allow the 7483 interpretation, but added a comment about them being mandatory in 9083.

File: rdap_types/src/lib.rs (hunk 7 above)
Comment (djc) [reply rdap_types/src/lib.rs:681]: Changed.

File: rdap_types/src/lib.rs (hunk 3 above)
Comment (djc) [reply rdap_types/src/lib.rs]: Yup, added an `Object` enum that provides the `objectClassName` field.

File: rdap_types/src/lib.rs (hunk 5 above)
Comment (djc) [reply rdap_types/src/lib.rs:148]: Replaced non-standard variants with an `Unknown(String)` variant.

**Review by valkum — APPROVED** [review]
Looks good.

**Pull request merged**